File: code/arduino/Encoder.cpp

/*
 * Class for Analog encoder for continous rotation servo motor', that inherits from Thread.
 * Put the Encoder.cpp file and Encoder.h files to include this as library for your projects.
 * Author: Maurice Tedder December 29, 2018
 */

#include "Encoder.h"

Encoder::Encoder(Servo servo, int analogSensor, int threshold, int servoMin, int servoMax, int servoStop){
    _servo = servo;
    _analogSensor = analogSensor;
    _threshold = threshold; 
    _servoMin = servoMin;
    _servoMax = servoMax;
    _servoStop = servoStop;
  }

  /*
   * Add command string to command queue
   */
  void Encoder::addCmd(String cmdStr){
    _queue.enqueue (cmdStr);
  }

//private:
//  Servo _servo; //servo controlled by this Encoder object
//  int _analogSensor; //analog sensor port
//  int _threshold; //analog sensor encoder wheel signal threshold
//  Queue <String> _queue; //command queue
//  bool _tick = false;
//  bool _tickp = false;
//  bool _continueLoop = true;
//  int _count = 0;
//  int _cmd = 0;
//  int _servoMin; //Minimum allowed servo motor pwm value
//  int _servoMax; //Maximum allowed servo motor pwm value
//  int _servoStop; //Stop servo motor pwm value

  // No, "run" cannot be anything...
  // Because Thread uses the method "run" to run threads,
  // we MUST overload this method here. using anything other
  // than "run" will not work properly...
  void Encoder::run(){
     // dequeue all the message's characters from the queue.
    while (!_queue.isEmpty()){
      _continueLoop = true;
      String data = _queue.dequeue();
      int delIndex = data.indexOf(',');
      //Serial.println (delIndex);
      int motorSpeed = (data.substring(0, delIndex)).toInt();//parse motor speed value
      _cmd = (data.substring(delIndex+1)).toInt();//parse position cmd value
      _servo.write(constrain(motorSpeed, _servoMin, _servoMax));
      
      while (_continueLoop) {//position control feedback loop
        int val = analogRead(_analogSensor); //get current value of wheel encoder  
        
        if (val < _threshold){//detect encoder wheel edge
          _tick = true;
        }else{
          _tick = false;
        }
                            
        if (_tick != _tickp){
          _count = _count + 1;
        }
      
        _tickp = _tick;
      
        if((_cmd - _count) < 0){//get next command after completion of previous command   
            _count = 0;
            _tick = false;
            _tickp = false;
            _continueLoop = false;
            _servo.write(constrain(_servoStop, _servoMin, _servoMax));
        }
                          
      }//end position control feedback loop
    }//end command queue read loop
  }//end run methods

File: code/raspberryPiZeroW/encoder.py
import queue
import threading
import automationhat
import smbus #I2C interface

# Wheel encoder code for analog sensors attached to a continous rotation servo motor.
#By Maurice Tedder
#December 25, 2018
#Ref: https://www.tutorialspoint.com/python/python_multithreading.htm

THRESHOLD = 2.5

bus = smbus.SMBus(1)
addr = 0x40

#init servo duty cycle
#bus.write_byte_data(addr, 0, 0x20) #200 Hz
#bus.write_byte_data(addr, 0xfe, 0x1e) #200 Hz
bus.write_byte_data(addr, 0, 0x20) #50 HZ
bus.write_byte_data(addr, 0xfe, 0x1e) #50 Hz
#ch0
servo0_start = 0x06
servo0_end = 0x08
#ch1
servo1_start = 0x0A
servo1_end = 0x0C
#ch2
servo2_start = 0x0E
servo2_end = 0x10

#create synchronization lock
queueLock = threading.Lock()

def motor(bus, addr, servo_start, servo_end, pwm):#activate servo hat servo motor       
    bus.write_word_data(addr, servo_start, 0)    
    bus.write_word_data(addr, servo_end, pwm)
        
        
class Encoder(threading.Thread):
    'Class for Analog encoder on continous rotation servo motors'
  
    def __init__(self, bus, addr, servo_start, servo_end, threshold):
        threading.Thread.__init__(self)
        self.q = queue.Queue() #q, command queue
        self.encoderLoop = True
        self.bus = bus
        self.addr = addr
        self.servo_start = servo_start
        self.servo_end = servo_end
        self.threshold = threshold #analog sensor encoder wheel signal threshold

    def startEncoder(self):#start encoder loop
            self.encoderLoop = True
            
    def stopEncoder(self):#Stop encoder loop
            self.encoderLoop = False
        
    def addCmd(self, cmd):#function to add command tuple to the command queue
        queueLock.acquire()
        self.q.put(cmd)
        queueLock.release()
        
    def run(self):        
        tick = False
        tickp = False
        count = 0
        continueLoop = False #encoder tick counter loop flag
        while self.encoderLoop:
            queueLock.acquire()            
            if not self.q.empty():#Test for new commands in the queue
                continueLoop = True
                data = self.q.get() #get command tuple from queue and parse data
                pwm = data[0]
                cmd = data[1]                       
                motor(self.bus, self.addr, self.servo_start, self.servo_end, pwm)
                queueLock.release()
            else:
                queueLock.release()                
                    
            while continueLoop:#position control feedback loop
                #print(str(self.threshold))
                val = automationhat.analog[0].read() #get current value of wheel encoder sensor                
                if (val < self.threshold):
                    tick = True
                else:
                    tick = False
                    
                if (tick != tickp):
                    count = count + 1

                tickp = tick

                if((cmd - count) < 0):#get next command after completion of previous command                    
                    count = 0
                    tick = False
                    tickp = False
                    continueLoop = False                                                        
    
#create motor command list
cmds = [(1639, 20), (1250, 0),(1000, 50), (1250, 0)]
#cmds = [(1639, 0)]

#Create encoder object
e1 = Encoder(bus, addr, servo0_start, servo0_end,THRESHOLD)

#add commands from command list to encoder command queue
for dat in cmds:
    e1.addCmd(dat)

#start encoder thread
e1.start()


#Uncomment this section to use interactive servo command mode
#--- Start interactive mode servo command ---
##cmds = []
##doLoop = True
##while doLoop:
##    speed = int(input("Input Speed:"))
##    cmd = int(input("Input command ticks:"))
##    cmds.append((speed, cmd))
##    
##    user_input = input("Would you like send your commands? enter (y = yes):")
##    if user_input == 'y':
##        for dat in cmds:
##            e1.addCmd(dat)
##        cmds = [] #reset commands array
##        
##    user_input = input("Would you like to continue? enter (y = yes):")
##    if user_input != 'y':
##        doLoop = False
#        
#---End interactive servo command mode code ---
#

#wait for thread to complete
#e1.join()

#end encoder thread
user_input = input("Enter y to exit:")
if user_input == 'y':
    e1.stopEncoder() #end encoder thread
    
#wait for thread to complete
e1.join()

print("THE END")
